### PR TITLE
Various module import fixes

### LINKED ISSUE #3257: Module imports in ch.exe/jshost.exe should be relative to the current module
Currently, all imports are relative to CWD. It should instead be relative to the module doing the import.

### LINKED ISSUE #5237: Module imports in ch should be relative to the current module
For comparison of esmodule behavior to any other environment imports should really be relative to the current module. (e.g. to investigate #5171 I had to use @fatcerberus miniSphere as my test harness to be able to try and load the ESM version of moment.js as it assumed relative imports so could not test in ch)

Currently in ch all imports are relative to the CWD instead.

I note that there was a previous issue about this #3257 which was fixed BUT since then the behaviour has been reverted.

There was a test case to confirm that imports were relative to current module not CWD https://github.com/Microsoft/ChakraCore/blob/master/test/es6module/bug_issue_3257.js But after it was broken a few times it was re-enabled without paths and hence no longer tested the original issue. (Though it remains a very useful test of circular imports combined with dynamic import)

### LINKED ISSUE #5795: BUG: Dynamic import should return different promise objects, even if the specifier is the same.
Test: https://github.com/tc39/test262/blob/master/test/language/expressions/dynamic-import/always-create-new-promise.js

Result: 

> the returned promises are not the same, regardless the reference and specifier pair Expected SameValue(«[object Promise]», «[object Promise]») to be false

### LINKED ISSUE #5796: BUG: Dynamic import should only return a rejected promise if it fails loading a module
Test: https://github.com/tc39/test262/blob/master/test/language/expressions/dynamic-import/assignment-expression/unary-expr.js

Result: 

> Couldn't load file 'true'
> Couldn't load file 'undefined'
> Couldn't load file 'object'
> Couldn't load file 'NaN'
> Couldn't load file '-1'

Expected: every import call should return a promise that will be resolved/rejected at some later time. Presently, the import fails and produces these errors synchronously.

### LINKED ISSUE #5797: BUG: ImportCall is a CallExpression, it can't be preceded by the new keyword
Test: https://github.com/tc39/test262/blob/master/test/language/expressions/dynamic-import/syntax/invalid/top-level-no-new-call-expression.js

TL;DR:

```js
new import('');
```

Is not allowed, per: https://tc39.github.io/proposal-dynamic-import/#sec-left-hand-side-expressions

### PR DESCRIPTION
**Pull request by rhuanjl**
This PR fixes a collection of module related issues - some of these are very small so I thought better to combine them into one PR. I hope this isn't too much at once.

1. Restore the use of relative paths for imports within modules in ch - this was implemented a long time ago and then broken - the functionality was mostly still there just broken in a few places - originally this was issue 3257, it also has a newer issue I opened 5237. As well as restoring the functionality I restored the test case from 3257 which was no longer testing what it had been designed for after a previous edit.

2. Don't print messages about failing to load modules or syntax errors in modules when processing dynamic imports - this was purely a ch issue but was problematic for several test262 cases per issue 5796.

3. Reject the promise from import() if a child module of the dynamically imported module throws a runtime error - there was a bug where this promise was never being resolved - this is also part of 5796.

4. `new import(anything)` should be a syntax error - 5797

5. import() should return a different promise each time even when using the same specifier - fix this by using JavascriptPromise::CreatePassThroughPromise 5795

cc @rwaldron @boingoing 

fix #5796
fix #5795
fix #5797 
fix #5237 
re-fix #3257

### CODE REVIEW

**Comment by rhuanjl** [comment]
Closing temporarily whilst I dig into some windows only test failures.

File: bin/ch/WScriptJsrt.cpp [hunk 1]
```diff
@@ -47,6 +47,7 @@
 MessageQueue* WScriptJsrt::messageQueue = nullptr;
 std::map<std::string, JsModuleRecord>  WScriptJsrt::moduleRecordMap;
 std::map<JsModuleRecord, std::string>  WScriptJsrt::moduleDirMap;
+std::map<JsModuleRecord, int>  WScriptJsrt::moduleErrMap;
```
Comment (boingoing) [inline bin/ch/WScriptJsrt.cpp]: Can we replace int here with an enum? I'd like to avoid the use of magic numbers.

File: bin/ch/WScriptJsrt.h [hunk 2]
```diff
@@ -140,5 +141,6 @@ class WScriptJsrt
     static DWORD_PTR sourceContext;
     static std::map<std::string, JsModuleRecord> moduleRecordMap;
     static std::map<JsModuleRecord, std::string> moduleDirMap;
+    static std::map<JsModuleRecord, int> moduleErrMap; // 0 = root module without error reported, 1 = fetched module, 2 = module with error reported
```
Comment (boingoing) [inline bin/ch/WScriptJsrt.h]: See above, let's please use an enum with named values instead of magic numbers.

File: lib/Parser/Parse.cpp [hunk 3]
```diff
@@ -3624,6 +3624,10 @@ ParseNodePtr Parser::ParseTerm(BOOL fAllowCall,
     case tkIMPORT:
         if (m_scriptContext->GetConfig()->IsES6ModuleEnabled() && m_scriptContext->GetConfig()->IsESDynamicImportEnabled())
         {
+            if (!fAllowCall)
+            {
+                Error(ERRTokenAfter, _u("import"), _u("new"));
```
Comment (boingoing) [inline lib/Parser/Parse.cpp]: Is this supposed to be a SyntaxError? I would expect this is a runtime error but (full disclosure) I didn't look closely at the spec.

File: lib/Runtime/Language/SourceTextModuleRecord.cpp [hunk 4]
```diff
@@ -342,7 +342,7 @@ namespace Js
             }
         }
 
-        return this->promise;
+        return JavascriptPromise::CreatePassThroughPromise(this->promise, scriptContext);
```
Comment (boingoing) [inline lib/Runtime/Language/SourceTextModuleRecord.cpp]: Nice 👍

File: lib/Parser/Parse.cpp (hunk 3 above)
Comment (rhuanjl) [reply lib/Parser/Parse.cpp]: Syntax error per test262: https://github.com/tc39/test262/blob/master/test/language/expressions/dynamic-import/syntax/invalid/top-level-no-new-call-expression.js

Comment (boingoing) [reply lib/Parser/Parse.cpp]: Seems right. I should have taken a look at the proposal. Appears dynamic import is a special type of CallExpression and can't be a NewExpression. Similar to SuperCall (which also throws SyntaxError when called with new). At least, this is according to https://tc39.github.io/proposal-dynamic-import/#prod-ImportCall

Thanks for confirming.

**Comment by rhuanjl** [comment]
Thanks for the review.

There are still a few more issues to solve 3251/5236, 5778 and 5798. I don't have fixes for any of those (yet) though.

**Comment by boingoing** [comment]
Would you mind rebasing this one more time ?

File: test/es6module/rlexe.xml [hunk 5]
```diff
@@ -69,6 +69,12 @@
       <tags>exclude_sanitize_address</tags>
     </default>
   </test>
+  <test>
+    <default>
+      <files>dynamic_import_promises_5796.js</files>
+      <compile-flags>-ESDynamicImport</compile-flags>
```
Comment (boingoing) [inline test/es6module/rlexe.xml]: Would you mind adding `<tags>exclude_jshost</tags>` to this test and also the test for 3257? These are failing with our internal test host. I can update our host later but I don't see a reason to block your PR for it.

Comment (rhuanjl) [reply test/es6module/rlexe.xml]: done

Comment (boingoing) [reply test/es6module/rlexe.xml]: Thank you, I'll merge this in once the CI passes.

File: lib/Parser/Parse.cpp (hunk 3 above)
Comment (fatcerberus) [reply lib/Parser/Parse.cpp]: For what it’s worth I was surprised this was a syntax error too.  `import(“”) += 1` is *not* a syntax error (it’s a runtime error though), but this is.  The spec is weird sometimes.

Comment (rhuanjl) [reply lib/Parser/Parse.cpp]: import('') +=1 is meant to be an early error, though a Reference error not syntax error but it's not at runtime. Note currently CC does a Syntax error there's an open issue for it.

**Comment by rhuanjl** [comment]
Thank you for merging this @boingoing
